src/core/qgsvectorlayereditbuffer.cpp

```diff
@@ -310,6 +310,36 @@ bool QgsVectorLayerEditBuffer::commitChanges( QStringList &commitErrors )
   // no                 yes                   => changeAttributeValues
   // yes                yes                   => changeFeatures
 
+  // to fix https://issues.qgis.org/issues/15741
+  // first of all check if feature to add is compatible with provider type
+  // this check have to be done before all checks to avoid to clear internal
+  // buffer if some of next steps success.
+  if ( success && !mAddedFeatures.isEmpty() )
+  {
+    if ( cap & QgsVectorDataProvider::AddFeatures )
+    {
+      for ( QgsFeature f : mAddedFeatures )
+      {
+        if ( ( ! f.hasGeometry() ) ||
+             ( f.geometry().wkbType() == provider->wkbType() ) )
```

Comment:
Do we need the check here? Is it not sufficient the check at line 328?
I mean that according to @mhugent advice, the conversion should be better handled on the provider, that might be able to convert the type, so, instead of checking for the type match, I would prefer just trying delegate the provider to make the conversion and fail if the provider is not capable.